File: backend/src/main/java/service/UniversityService.java
package service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import javax.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import dto.mapper.ModelMapper;
import dto.model.UniversityDto;
import exception.UniversityException.UniversityNotFoundException;
import exception.CountryException.CountryNullException;
import model.Country;
import model.University;
import repository.UniversityRepository;

@Service
@Transactional
public class UniversityService {

    @Autowired
    private UniversityRepository repository;

    @Autowired
    private CountryService countryService;

    public University findById(Long id) {
        Optional<University> itemOp = repository.findById(id);

        if (!itemOp.isPresent()) {
            throw new UniversityNotFoundException(id);
        }

        return itemOp.get();
    }

    public List<University> findAll() {
        List<University> items = new ArrayList<>();

        for (University item : repository.findAll()) {
            items.add(item);
        }

        return items;
    }

    public UniversityDto create(UniversityDto itemDto) {
        University item = new University();
        item.setName(itemDto.getName());
        item.setAcronym(itemDto.getAcronym());
        item.setUrl(itemDto.getUrl());

        if (itemDto.getCountryId() == null) {
            throw new CountryNullException();
        }
        
        Country country = countryService.findById(itemDto.getCountryId());
        item.setCountry(country);

        return ModelMapper.toUniversityDto(repository.save(item));
    }

    public UniversityDto read(Long id) {
        University item = findById(id);
        return ModelMapper.toUniversityDto(item);
    }

    public List<UniversityDto> readAll() {
        List<UniversityDto> items = new ArrayList<>();

        for (University item : repository.findAll()) {
            items.add(ModelMapper.toUniversityDto(item));
        }

        return items;
    }

    public List<UniversityDto> readByCountry(Long countryId) {
        List<UniversityDto> items = new ArrayList<>();

        for (University item : repository.findByCountryId(countryId)) {
            items.add(ModelMapper.toUniversityDto(item));
        }

        return items;
    }

    public UniversityDto update(UniversityDto newItem, Long id) {
        University item = findById(id);

        if (newItem.getName() != null) {
            item.setName(newItem.getName());
        }

        if (newItem.getAcronym() != null) {
            item.setAcronym(newItem.getAcronym());
        }

        if (newItem.getUrl() != null) {
            item.setUrl(newItem.getUrl());
        }

        if (newItem.getCountryId() != null) {
            Country country = countryService.findById(newItem.getCountryId());
            item.setCountry(country);
        }

        return ModelMapper.toUniversityDto(repository.save(item));
    }

    public void delete(Long id) {
        University item = findById(id);
        repository.delete(item);;
    }

}

File: frontend/src/app/@core/services/university.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from './../../../environments/environment';

@Injectable()
export class UniversityService {

  static path = environment.APIEndpoint + '/university/';

  constructor(private http: HttpClient) { }

  create(item: any): any {
    const path = UniversityService.path;
    return this.http.post(path, item).toPromise();
  }

  read(id: number): any {
    const path = UniversityService.path;
    return this.http.get(path + id.toString()).toPromise();
  }

  readAll(): any {
    const path = UniversityService.path;
    return this.http.get(path).toPromise();
  }

  readByCountry(id: number): any {
    const path = UniversityService.path;
    return this.http.get(path + 'country/' + id.toString() ).toPromise();
  }

  update(id: number, item: any): any {
    const path = UniversityService.path;
    return this.http.put(path + id.toString(), item).toPromise();
  }

  delete(id: number): any {
    const path = UniversityService.path;
    return this.http.delete(path + id.toString()).toPromise();
  }

}
